### PR TITLE
fix: cleanup data after deleting a component

### LINKED ISSUE #20931: Deleting or renaming a component does not delete/update the relational linking data for the component
### Required System information

- Node.js version: node/v20.11.1 linux x64
- NPM version: yarn/1.22.22 npm/? 
- Strapi version: 4.25.6
- Database: SQLite
- Operating system: Linux Mint 21.3
- Is your project Javascript or Typescript: Javascript

### Describe the bug

When a component is deleted or rename the related component data is not properly deleted or updated (updating may not be easily possible).

In certain cases or certain database types this can lead to an error similar to:

```
error: Metadata for "category.component" not found
```

I was not able to reproduce this error but I can see the remaining artifacts within the database

### Steps to reproduce the behavior

1. Create a content-type containing several components or dynamic zones with components
2. Create some test data on the content-type including some component data
3. Delete or rename the component (renaming requires doing it manually through the schema files)
4. Check the database component linking tables (using something like dbeaver)
5. See that the component data still exists

### Expected behavior

When a component is deleted all references in the database to that component should also be deleted. In the case of renaming we may not be able to do that automatically until we refactor the CTB (similar to renaming a normal field)

### Screenshots

Content Structure prior to modification:
image

Example entry containing the data:
image

Database Prior to deleting the component:
image

Deleting a component and the data within the database:
image

### Code snippets

N/A

### Additional context

Related to TID7662

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

Cleans up the data in the database properly after deleting a component

TODO in follow-up PRs:

- [ ] add a migration and/or database repair script to remove this data from databases that are already affected
- [ ] [Graceful renaming/deletion of everything](https://feedback.strapi.io/developer-experience/p/gracefully-handle-renaming-of-content-types-and-fields-in-the-ctb)

### Why is it needed?

We aren't cleaning up all the data in the database after deleting a component, which also can cause issues when components with the same name are recreated

### How to test it?

Tests are included; see linked issue to test manually

### Related issue(s)/PR(s)

Mitigates #20931 for v5 -- full solution for affected production systems will be much more complex

DX-1726

### CODE REVIEW

File: packages/core/content-type-builder/server/src/services/components.ts [hunk 1]
```diff
@@ -86,13 +86,38 @@ export const editComponent = async (
...
+        (attr: any) => attr.target === deleteUid && attr.joinTable?.name
+      );
+
+      // Delete entries in each join table associated with matching attributes
+      for (const attr of matchingAttributes) {
+        if (!('joinTable' in attr && attr.joinTable && attr.joinTable.name)) {
+          continue;
+        }
+
+        await trx(attr.joinTable.name).where('component_type', deleteUid).delete();
```
Comment (innerdvations) [inline packages/core/content-type-builder/server/src/services/components.ts:109]: Any opinions on if this should be allowed to throw, or if this should be treated as a non-critical part of the delete and just warn? I already have a stashed change that just logs a warning that the data couldn't be deleted if there is a problem, which would make this change a lot safer in the case that we've missed any edge cases where we target the wrong table for example.

Comment (alexandrebodin) [reply packages/core/content-type-builder/server/src/services/components.ts:109]: I think this will rarely be usefull. the CTB is used locally so most likely not connected to a production database and as such will not change anything when the deployment happen. Considering this I would reco a warning

Comment (innerdvations) [reply packages/core/content-type-builder/server/src/services/components.ts:109]: True, this will just be preventing immediate errors. The migration or repair will be needed to fix it in production.

**Comment by jhoward1994** [comment]
```
  if (failedDeletes.length > 0) {
    strapi.log.warn(
      `Failed to delete component data for ${deleteUid} from ${failedDeletes.map((attr) => attr.table)}`
    );
  }
```

Should we make this warning message clearer that the component itself has been deleted but some references could be left in the DB. I'm just thinking that the way it's currently written could be ambiguous.

e.g. `Component '${deleteUid}' was deleted, but references remain in the following database tables: ${failedDeletes.map((attr) => attr.table)}. Consider cleaning up these tables to avoid orphaned data.`

**Review by butcherZ — APPROVED** [review]
QAed with `v0.0.0-experimental.702fcf7aec9be82c4485a8bd49e235293c37f022`

- created a component, added the component to a content type, created test data with the content type, delete the component
1.  test data is gone ✅
2. relations in database is gone ✅
3. content type is still available ✅
4. able to re-create component with the same name ✅

I have also tested the edge case mentioned in DX-1726
1. prod server starts up  ✅
2. no errors in CTB  ✅
3. no errors in component editor  ✅
4. no problems adding / editing those components  ✅

considering the edge case is a bit tricky, better to have someone to confirm with me on it :D

File: packages/core/content-type-builder/server/src/services/components.ts [hunk 2]
```diff
@@ -86,13 +86,60 @@ export const editComponent = async (
   return updatedComponent;
 };
 
-export const deleteComponent = async (uid: Internal.UID.Component) => {
-  const builder = createBuilder();
+export const deleteComponent = async (deleteUid: Internal.UID.Component) => {
+  const failedDeletes: any[] = [];
```
Comment (Convly) [inline packages/core/content-type-builder/server/src/services/components.ts:90]: Can we have a real type here pls :)

File: packages/core/content-type-builder/server/src/services/components.ts [hunk 3]
```diff
@@ -86,13 +86,60 @@ export const editComponent = async (
...
+    const builder = createBuilder();
+
+    const models = [...builder.contentTypes.entries(), ...builder.components.entries()];
+
+    // Find models that include an attribute with this component as a target and delete references to it from the db
+    for (const [modelUid] of models) {
+      const metadata = strapi.db.metadata.get(modelUid);
+
+      const matchingAttributes = Object.values(metadata.attributes).filter(
+        (attr: any) => attr.target === deleteUid && attr.joinTable?.name
```
Comment (Convly) [inline packages/core/content-type-builder/server/src/services/components.ts:102]: Just for good measure, should we also check that it is a relation? (also, don't we have an attribute type that we can use that is exported from the DB? If not, not a big deal)

**Comment by innerdvations** [comment]
Closing in favor of https://github.com/strapi/strapi/pull/22267
